packages/shared/src/util/features.ts

```diff
@@ -175,62 +175,131 @@ const areRulesOneSided = (
+  const visit = ({
+    feature,
+    linkedExperiments,
+  }: {
+    feature: FeatureInterface;
+    linkedExperiments?: ExperimentInterfaceStringDates[];
+  }): { stale: boolean; reason?: StaleFeatureReason } => {
+    if (visitedFeatures.has(feature.id)) {
+      return { stale: false };
     }
+    visitedFeatures.add(feature.id);
 
-    const linkedExperimentIds = linkedExperiments.map((e) => e.id);
-    if (
-      !linkedExperimentIds.every((id) =>
-        feature.linkedExperiments?.includes(id)
-      )
-    ) {
-      // eslint-disable-next-line no-console
-      console.error("isFeatureStale: linkedExperiments id mismatch");
-      return { stale: false, reason: "error" };
-    }
+    try {
+      if (feature.neverStale) return { stale: false };
+
+      if (!linkedExperiments) {
+        // If we don't have linkedExperiments in the context (e.g. evaluating dependents), use the baked array
+        linkedExperiments = (feature?.linkedExperiments ?? [])
+          .map((id) => experiments.find((e) => e.id === id))
+          .filter(Boolean) as ExperimentInterfaceStringDates[];
+      }
+      if (feature.linkedExperiments?.length !== linkedExperiments.length) {
+        // eslint-disable-next-line no-console
+        console.error("isFeatureStale: linkedExperiments length mismatch");
+        return { stale: false, reason: "error" };
+      }
+      const linkedExperimentIds: string[] | undefined = linkedExperiments
+        ? linkedExperiments.map((e) => e.id)
+        : undefined;
+      if (
+        linkedExperimentIds &&
+        !linkedExperimentIds.every((id) =>
+          feature.linkedExperiments?.includes(id)
+        )
+      ) {
+        // eslint-disable-next-line no-console
+        console.error("isFeatureStale: linkedExperiments id mismatch");
+        return { stale: false, reason: "error" };
+      }
 
-    const twoWeeksAgo = subWeeks(new Date(), 2);
-    const dateUpdated = getValidDate(feature.dateUpdated);
-    const stale = dateUpdated < twoWeeksAgo;
+      const twoWeeksAgo = subWeeks(new Date(), 2);
+      const dateUpdated = getValidDate(feature.dateUpdated);
+      const stale = dateUpdated < twoWeeksAgo;
 
-    if (!stale) return { stale };
+      if (!stale) return { stale };
 
-    // features with draft revisions are not stale
-    if (feature.hasDrafts) return { stale: false };
+      // features with draft revisions are not stale
+      if (feature.hasDrafts) return { stale: false };
 
-    const envSettings = Object.values(feature.environmentSettings ?? {});
+      // features with fresh dependents are not stale
+      if (features && features.length > 1) {
+        const dependentFeatures = getDependentFeatures(
+          feature,
+          features,
+          environments
+        );
+        const hasNonStaleDependentFeatures = dependentFeatures.some((id) => {
+          const f = features?.find((f) => f.id === id);
+          if (!f) return true;
+          return !visit({ feature: f }).stale;
```

Comment:
Is the depth of a feature tree limited to 1?

Reply:
no, it's unbounded

Reply:
Ok, I may be misunderstanding the flow here, but if it's unbounded would we not want to pass in `features`and `experiments` to this `visit` call?

Reply:
they're global to the `isFeatureStale()` scope so they can be accessed by each `visit()` iteration. They don't need to mutate for each visit.